src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.Android.cs

```diff
@@ -0,0 +1,431 @@
+                    _ids![i] = new string(p);
+                    _lengths![i] = entry.length;
+
+                    if (_lengths![i] < sizeof(AndroidTzDataHeader))
+                    {
+                        throw new InvalidOperationException(SR.InvalidOperation_BadIndexLength);
+                    }
+                }
+            }
+
+            private unsafe AndroidTzDataEntry LoadEntryAt(Stream fs, long position)
+            {
+                int size = sizeof(AndroidTzDataEntry);
+                Span<byte> entryBuffer = stackalloc byte[size];
+
+                fs.Position = position;
+
+                int bytesRead = 0;
+                int bytesLeft = size;
+
+                while (bytesLeft > 0)
+                {
+                    int b = fs.Read(entryBuffer);
+                    if (b == 0)
+                    {
+                        break;
+                    }
+
+                    bytesRead += b;
+                    bytesLeft -= b;
+                }
+
+                if (bytesLeft != 0)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.InvalidOperation_ReadTZError, _tzFilePath, position, size, bytesRead, size));
+                }
+
+                AndroidTzDataEntry entry;
+                for (int i = 0; i < 40; i++)
+                {
+                    entry.id[i] = entryBuffer[i];
+                }
+                entry.byteOffset = TZif_ToInt32(entryBuffer.Slice(40, 4));
+                entry.length = TZif_ToInt32(entryBuffer.Slice(44, 4));
+                entry.unused = TZif_ToInt32(entryBuffer.Slice(48));
+
+                return entry;
+            }
+
+            public string[] GetTimeZoneIds()
+            {
+                return _ids!;
+            }
+
+            // On Android, time zone data is found in tzdata
+            // Based on https://github.com/mono/mono/blob/main/mcs/class/corlib/System/TimeZoneInfo.Android.cs
+            // Also follows the locations found at the bottom of https://github.com/aosp-mirror/platform_bionic/blob/master/libc/tzcode/bionic.cpp
+            public string GetTimeZoneDirectory()
+            {
+                return _tzFilePath!;
+            }
+
+            public byte[] GetTimeZoneData(string id)
+            {
+                int i = Array.BinarySearch(_ids!, id, StringComparer.Ordinal);
+                if (i < 0)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.TimeZoneNotFound_MissingData, id));
+                }
+
+                int offset = _byteOffsets![i];
+                int length = _lengths![i];
+                var buffer = new byte[length];
+                using (FileStream fs = File.OpenRead(_tzFilePath))
+                {
+                    fs.Position = offset;
+                    int numBytesRead;
+                    if ((numBytesRead = fs.Read(buffer)) < buffer.Length)
```

Comment:
Read in a loop until the buffer is full.

Reply:
Thanks, missed this one!

Reply:
Would it be helpful to have a common read the buffer method since we're doing it in a few places?

Reply:
Thanks, created another method that can be used to create a `Span<byte>` given a `FileStream`, `position`, and `size` of the `Span`.